src/drivers/smart_battery/batmon/batmon.cpp

```diff
@@ -0,0 +1,352 @@
+
+	new_report.average_current_a = average_current;
+
+	// Read run time to empty (minutes).
+	ret |= _interface->read_word(BATT_SMBUS_RUN_TIME_TO_EMPTY, result);
+	new_report.run_time_to_empty = result;
+
+	// Read average time to empty (minutes).
+	ret |= _interface->read_word(BATT_SMBUS_AVERAGE_TIME_TO_EMPTY, result);
+	new_report.average_time_to_empty = result;
+
+	// Read remaining capacity.
+	ret |= _interface->read_word(BATT_SMBUS_REMAINING_CAPACITY, result);
+
+	// Calculate total discharged amount in mah.
+	new_report.discharged_mah = _batt_startup_capacity - (float)result;
+
+	// Read Relative SOC.
+	ret |= _interface->read_word(BATT_SMBUS_RELATIVE_SOC, result);
+
+	// Normalize 0.0 to 1.0
+	new_report.remaining = (float)result / 100.0f;
+
+	// Read Max Error
+	//ret |= _interface->read_word(BATT_SMBUS_MAX_ERROR, result); //TODO: to be implemented
+	//new_report.max_error = result;
+
+	// Read battery temperature and covert to Celsius.
+	ret |= _interface->read_word(BATT_SMBUS_TEMP, result);
+	new_report.temperature = ((float)result / 10.0f) + CONSTANTS_ABSOLUTE_NULL_CELSIUS;
+
+	// Only publish if no errors.
+	if (ret == PX4_OK) {
+		new_report.capacity = _batt_capacity;
+		new_report.cycle_count = _cycle_count;
+		new_report.serial_number = _serial_number;
+		new_report.max_cell_voltage_delta = _max_cell_voltage_delta;
+		new_report.cell_count = _cell_count;
+		new_report.state_of_health = _state_of_health;
+
+		// TODO: This critical setting should be set with BMS info or through a paramter
+		// Setting a hard coded BATT_CELL_VOLTAGE_THRESHOLD_FAILED may not be appropriate
+		//if (_lifetime_max_delta_cell_voltage > BATT_CELL_VOLTAGE_THRESHOLD_FAILED) {
+		//	new_report.warning = battery_status_s::BATTERY_WARNING_CRITICAL;
+
+		if (new_report.remaining > _low_thr) {
+			new_report.warning = battery_status_s::BATTERY_WARNING_NONE;
+
+		} else if (new_report.remaining > _crit_thr) {
+			new_report.warning = battery_status_s::BATTERY_WARNING_LOW;
+
+		} else if (new_report.remaining > _emergency_thr) {
+			new_report.warning = battery_status_s::BATTERY_WARNING_CRITICAL;
+
+		} else {
+			new_report.warning = battery_status_s::BATTERY_WARNING_EMERGENCY;
+		}
+
+		new_report.interface_error = perf_event_count(_interface->_interface_errors);
+
+		int instance = 0;
+		orb_publish_auto(ORB_ID(battery_status), &_batt_topic, &new_report, &instance);
+
+		_last_report = new_report;
+	}
+}
+
+int Batmon::get_batmon_startup_info()
+{
+	int ret = PX4_OK;
+
+	// Read battery threshold params on startup.
+	param_get(param_find("BAT_CRIT_THR"), &_crit_thr);
+	param_get(param_find("BAT_LOW_THR"), &_low_thr);
+	param_get(param_find("BAT_EMERGEN_THR"), &_emergency_thr);
+
+	// Read BatMon specific data
+	uint16_t num_cells;
+	ret = _interface->read_word(BATT_SMBUS_CELL_COUNT, num_cells);
+	_cell_count = math::min((uint8_t)num_cells, (uint8_t)MAX_CELL_COUNT);
+
+	int32_t _num_cells = num_cells;
+	param_set(param_find("BAT_N_CELLS"), &_num_cells);
```

Comment:
I don't think this needs to change the parameter. It is a deprecated parameter and only read by battery drivers anyways

Reply:
1. Will change it to BAT1_N_CELLS.
2. Is there a downside to updating BAT1_N_CELLS from driver? We are able to see that in the logs and use it to determine the cell count for flight.